### PR TITLE
Scala integration

### PR DESCRIPTION
**Pull request by ivonu**
Add support to generate Scala Code

### CODE REVIEW

File: Stainless/lib/scalaz3-mac-64-3.jar
Comment (maxime-esa) [inline Stainless/lib/scalaz3-mac-64-3.jar:1]: Adding more than 100MB of binary files including .exe in the repo is not a good idea. These are not dependencies of the tool itself. Can you download them from an external source rather, when they are needed?

**Review by maxime-esa — COMMENTED** [review]
please check the few comments, in particular related to the inclusion of binary files

File: BackendAst/DAstUtilFunctions.fs [hunk 1]
```diff
@@ -12,17 +12,17 @@ open Language
 
 
 
-let getAccessFromScopeNodeList (ReferenceToType nodes)  (childTypeIsString: bool) (lm:LanguageMacros) (pVal : CallerScope) =
+let getAccesssFromScopeNodeList (ReferenceToType nodes)  (childTypeIsString: bool) (lm:LanguageMacros) (pVal : CallerScope) =
```
[the remark this answers is not in the document]
Comment (ivonu) [reply BackendAst/DAstUtilFunctions.fs]: fixed

File: BackendAst/DAstUtilFunctions.fs [hunk 2]
```diff
@@ -12,17 +12,17 @@ open Language
...
     let handleNode zeroBasedSeqeuenceOfLevel (pVal : CallerScope) (n:ScopeNode) (childTypeIsString: bool) = 
         match n with
         | MD _
         | TA _
         | PRM _
-        | VA _              -> raise(BugErrorException "getAccessFromScopeNodeList")
+        | VA _              -> raise(BugErrorException "getAccesssFromScopeNodeList")
```
[the remark this answers is not in the document]
Comment (ivonu) [reply BackendAst/DAstUtilFunctions.fs]: fixed

File: BackendAst/DAstUtilFunctions.fs [hunk 3]
```diff
@@ -40,7 +40,7 @@ let getAccessFromScopeNodeList (ReferenceToType nodes)  (childTypeIsString: bool
                 let zeroBasedSeqeuenceOfLevel = match n with SQF -> zeroBasedSeqeuenceOfLevel + 1 | _ -> zeroBasedSeqeuenceOfLevel
                 (newPath, chekPath@curCheckExp, zeroBasedSeqeuenceOfLevel, idx+1)) (pVal,[], 0, 1) |> (fun (a,chekPath,_,_) -> a, chekPath)
         ret 
-    | _                                 -> raise(BugErrorException "getAccessFromScopeNodeList")
+    | _                                 -> raise(BugErrorException "getAccesssFromScopeNodeList")
```
[the remark this answers is not in the document]
Comment (ivonu) [reply BackendAst/DAstUtilFunctions.fs]: fixed

File: BackendAst/DAstUtilFunctions.fs [hunk 4]
```diff
@@ -175,6 +177,7 @@ type ChChildInfo with
     member this.presentWhenName (defOrRef:TypeDefintionOrReference option) l = 
         match l with
         | C     -> (ToC this._present_when_name_private) + "_PRESENT"
+        | Scala -> (ToC this._present_when_name_private) + "_PRESENT" // TODO: Scala
```
[the remark this answers is not in the document]
Comment (ivonu) [reply BackendAst/DAstUtilFunctions.fs:180]: While adding the Scala backend, we noticed that in some cases the Ada and C cases are different.
So we added some TODOs there, as these might need to be modified.

File: Stainless/lib/scalaz3-mac-64-3.jar
Comment (ivonu) [reply Stainless/lib/scalaz3-mac-64-3.jar:1]: you are right. Those binaries are not needed yet.
Later on, we need to think of a way to run Stainless to verify the Scala code.

**Pull request merged**